lib/jsonapi_parameters/translator.rb

```diff
@@ -16,15 +16,26 @@ def jsonapi_translate(params, naming_convention:)
 
     return params if params.nil? || params.empty?
 
-    @jsonapi_unsafe_hash = if naming_convention != :snake || JsonApi::Parameters.ensure_underscore_translation
-                             params = params.deep_transform_keys { |key| key.to_s.underscore.to_sym }
-                             params[:data][:type] = params[:data][:type].underscore if params.dig(:data, :type)
-                             params
-                           else
-                             params.deep_symbolize_keys
-                           end
+    @jsonapi_unsafe_hash = ensure_naming(params, naming_convention)
+
+    JsonApi::Parameters::Validator.new(@jsonapi_unsafe_hash.deep_dup).validate! if should_prevalidate?
```

Comment:
`validate!` has been provided by rails >= 5.0
So this solution limits the benefits for rails 4.0, which are not maintained anymore.
Michał Książek is coding solution to change that

Reply:
That should be pretty easy to overcome, as `#validate` is present since Rails 4.2.1: https://apidock.com/rails/v4.2.1/ActiveModel/Validations/validate

So perhaps it is worth to check whether this method is defined, and if not, then just call `validate` and raise on failure.

Reply:
Since we already limit ourselves to >= 4.1.8, we might as well limit ourselves to 4.2.1.

https://github.com/visualitypl/jsonapi_parameters/blob/master/jsonapi_parameters.gemspec#L25

Reply:
We spoke with Michał to drop activeModel validation at all - it does not provide big value

Reply:
Indeed from what I can see in the upcoming PR to this branch, this seems like a move in good direction. We just have to ensure this will be a solid solution, as this will be something technically "invented here" whereas up until this point we have used proven and well tested ActiveModel validations.